### PR TITLE
[featurebranch] Post adr UUID doctrine

### PR DESCRIPTION
**Pull request by Balamung**
This PR is only CPM-629: Set the identifier optional in mysql database.

### CODE REVIEW

**Review by tseho — APPROVED** [review]

**Review by pierallard — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/MigrateZddCommand.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        private LoggerInterface $logger,
+        \Traversable $zddMigrations
+    ) {
+        $this->zddMigrations = iterator_to_array($zddMigrations);
+
+        Assert::allIsInstanceOf($this->zddMigrations, ZddMigration::class);
+        usort($this->zddMigrations, fn ($a, $b) => \strcmp(
+            (new \ReflectionClass($a))->getShortName(),
+            (new \ReflectionClass($b))->getShortName()
+        ));
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/MigrateZddCommand.php:36]:
```suggestion
        iterable $zddMigrations
    ) {
        Assert::allIsInstanceOf($zddMigrations, ZddMigration::class);
        $zddMigrations = $zddMigrations instanceof \Traversable ? \iterator_to_array($zddMigrations) : $zddMigrations;
        Assert::uniqueValues(\array_map(static fn(ZddMigration $migration): string => $migration->getName(), $zddMigrations, 'Migration names should be unique'));

        \usort($zddMigrations, fn ($a, $b): int => (new \ReflectionClass($a))->getShortName() <=> (new \ReflectionClass($b))->getShortName()));
        $this->zddMigrations = $zddMigrations;
```

File: src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/MigrateZddCommand.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                    $this->logger->error(
+                        sprintf('pim:zdd-migration:migrate errored %s', $zddMigration->getName()),
+                        [
+                            'name' => $zddMigration->getName(),
+                            'exception' => $e,
+                        ]
+                    );
+
+                    $output->write($e->getMessage());
+                    return Command::FAILURE;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/MigrateZddCommand.php:82]:
```suggestion

                    return Command::FAILURE;
```

File: src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/ZddMigration.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Enrichment\Bundle\Command\ZddMigrations;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/ZddMigration.php:5]: probably not the good namespace, maybe rather somewhere in `Akeneo\Tool\Bundle\InstallerBundle` (not quite right either, but better IMO if we want to let other teams use this)

File: src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/MigrateZddCommand.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Webmozart\Assert\Assert;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license https://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class MigrateZddCommand extends Command
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/MigrateZddCommand.php:18]: final

File: src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/MigrateZddCommand.php [hunk 5]
```diff
@@ -0,0 +1,129 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Enrichment\Bundle\Command\ZddMigrations;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/MigrateZddCommand.php:5]: to move in another namespace?

File: src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/V20220516171405SetProductIdentifierNullableZddMigration.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+              AND column_name=:columnName;
+        SQL;
+
+        $result = $this->connection->fetchOne($sql, [
+            'schema' => $schema,
+            'tableName' => $tableName,
+            'columnName' => $columnName
+        ]);
+
+        return $result !== 'NO';
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/ZddMigrations/V20220516171405SetProductIdentifierNullableZddMigration.php:55]:
```suggestion
        return $result === 'YES';
```
(if the column does not exist anymore we don't want to execute the ALTER TABLE query)

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Db/MarkZddMigrationsAsMigratedSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        private Connection $connection,
+        \Traversable $zddMigrations
+    ) {
+        $this->zddMigrations = iterator_to_array($zddMigrations);
+
+        Assert::allIsInstanceOf($this->zddMigrations, ZddMigration::class);
+        usort($this->zddMigrations, fn ($a, $b) => \strcmp(
+            (new \ReflectionClass($a))->getShortName(),
+            (new \ReflectionClass($b))->getShortName()
+        ));
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Db/MarkZddMigrationsAsMigratedSubscriber.php:36]: no need to sort in this context

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Db/MarkZddMigrationsAsMigratedSubscriber.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    /** @var ZddMigration[] */
+    private array $zddMigrations;
+
+    public function __construct(
+        private Connection $connection,
+        \Traversable $zddMigrations
+    ) {
+        $this->zddMigrations = iterator_to_array($zddMigrations);
+
+        Assert::allIsInstanceOf($this->zddMigrations, ZddMigration::class);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Db/MarkZddMigrationsAsMigratedSubscriber.php:32]:
```suggestion
        iterable $zddMigrations
    ) {
        Assert::allIsInstanceOf($zddMigrations, ZddMigration::class);
        $zddMigrations = $zddMigrations instanceof \Traversable ? \iterator_to_array($zddMigrations) : $zddMigrations;
        $this->zddMigrations =$zddMigrations;
```

**Review by mmetayer — APPROVED** [review]

**Review by mmetayer — APPROVED** [review]
GTM once the namespace comment is resolved, other comments are small suggestions to improve maintainability in the future, up to you to take them into account or not

**Pull request merged**